avocado-core/src/main/scala/org/bdgenomics/avocado/genotyping/SquareOffReferenceModel.scala

```diff
@@ -0,0 +1,214 @@
+ * genotype likelihoods from the reference models.
+ */
+object SquareOffReferenceModel {
+
+  /**
+   * Squares off genotypes containing both called sites and reference models.
+   *
+   * @param genotypes Genotypes containing both called sites and reference models.
+   * @return A set of variant contexts where at least one copy of the alternate
+   *   allele was called across all samples, with genotype likelihood models for
+   *   all samples that had data at the site.
+   */
+  def apply(genotypes: GenotypeRDD): VariantContextRDD = {
+
+    val variants = extractVariants(genotypes)
+
+    // join variants back against genotypes
+    val sites = variants.shuffleRegionJoinAndGroupByLeft(genotypes)
+
+    sites.transmute(_.map(s => squareOffSite(s._1, s._2)))
+  }
+
+  /**
+   * Trims the rightmost end of a ref/alt allele pair.
+   *
+   * Needed to canonicalize alleles across samples. If the two alleles in a
+   * ref/alt allele pair end with the same suffix, trims those bases.
+   *
+   * @param ref The reference allele.
+   * @param alt The alternate allele.
+   * @return Returns the number of bases to trim from the two alleles, if any.
+   */
+  private[genotyping] def trimRight(ref: String,
+                                    alt: String): Int = {
+
+    @tailrec def trim(refIter: Iterator[Char],
+                      altIter: Iterator[Char],
+                      trimmed: Int = 0): Int = {
+      if (refIter.next != altIter.next) {
+        trimmed
+      } else if (!(refIter.hasNext && altIter.hasNext)) {
+        trimmed
+      } else {
+        trim(refIter, altIter, trimmed + 1)
+      }
+    }
+
+    if (ref.isEmpty || alt.isEmpty) {
+      0
+    } else {
+      trim(ref.reverse.toIterator,
+        alt.reverse.toIterator)
+    }
+  }
+
+  /**
+   * Discovers variant sites from the reference model genotypes.
+   *
+   * @param genotypes Genotypes containing both called sites and reference models.
+   * @return Returns sites where a variant was seen in at least one sample.
+   */
+  private[genotyping] def extractVariants(genotypes: GenotypeRDD): VariantRDD = {
+
+    genotypes.transmute(_.flatMap(gt => {
+
+      if (gt.getAlleles.contains(GenotypeAllele.ALT) &&
+        gt.getVariant.getAlternateAllele != null) {
+
+        // trim the alleles so that we have a canonical result
+        val toTrimFromRight = trimRight(gt.getVariant.getReferenceAllele,
+          gt.getVariant.getAlternateAllele)
+
+        Some(Variant.newBuilder
+          .setContigName(gt.getContigName)
+          .setStart(gt.getStart)
+          .setEnd(gt.getEnd - toTrimFromRight)
+          .setReferenceAllele(gt.getVariant
+            .getReferenceAllele
+            .dropRight(toTrimFromRight))
+          .setAlternateAllele(gt.getVariant
+            .getAlternateAllele
+            .dropRight(toTrimFromRight))
+          .build)
+      } else {
+        None
+      }
+    })).transformDataset(_.dropDuplicates("start",
```

Comment:
Do you want to add doc that this class realizes the RDD as a dataset, or is that expected across Avocado?

Reply:
There's a mix, but overall the codebase is trending towards SQL.